zaza/openstack/charm_tests/rabbitmq_server/tests.py

```diff
@@ -57,6 +58,41 @@ def _retry_get_amqp_message(self, check_unit, ssl=None, port=None):
                                                   ssl=ssl,
                                                   port=port)
 
+    @tenacity.retry(reraise=True, stop=tenacity.stop_after_attempt(30),
+                    retry=tenacity.retry_if_exception_type(AssertionError))
+    def _search_for_message(self, amqp_msg, check_unit, ssl, port,
+                            amqp_msg_counter):
+        """Search for message in message queue.
+
+        WARNING: This will consume messages until it finds the target message.
+
+        :param amqp_msg: Message to search for
+        :type amqp_msg: string
+        :param check_unit: Unit to retrieve messages from
+        :type check_unit: juju.unit.Unit
+        :param ssl: Whether to use SSL when connecting to rabbit
+        :type ssl: bool
+        :param port: Port to use when connecting to rabbit
+        :type port: Union[int, None]
+        :param amqp_msg_counter: Number in test sequence of this message.
+        :type amqp_msg: int
+        :raises: RmqNoMessageException
+        """
+        amqp_msg_rcvd = self._retry_get_amqp_message(
+            check_unit,
+            ssl=ssl,
+            port=port)
+
+        try:
+            logging.info('Looking for message {}'.format(amqp_msg))
+            # Validate amqp message content
+            assert amqp_msg == amqp_msg_rcvd
+            logging.info('Message {} received OK.'.format(amqp_msg_counter))
+        except AssertionError as err:
+            logging.info('Expected: {}'.format(amqp_msg))
+            logging.info('Actual:   {}'.format(amqp_msg_rcvd))
+            raise err
```

Comment:
Thanks, this is now indeed better but I still see this:

This line may raise an AssertionError 31 times, swallowed 30 times by tenacity.retry and then finally re-raised, correct? So then `_search_for_message()` can in theory raise an AssertionError too, correct?

If you agree, these two things should still be extended:
* rtype on line 79
* except type on line 162

Thanks!